Sources/AWSSDKSwiftCore/HTTP/NIOTSHTTPClient.swift

```diff
@@ -0,0 +1,225 @@
+
+/// HTTP Client class providing API for sending HTTP requests
+@available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *)
+public final class NIOTSHTTPClient {
+
+    /// Request structure to send
+    public struct Request {
+        var head: HTTPRequestHead
+        var body: ByteBuffer?
+    }
+
+    /// Response structure received back
+    public struct Response {
+        let head: HTTPResponseHead
+        let body: ByteBuffer?
+    }
+
+    /// Errors returned from HTTPClient when parsing responses
+    public enum HTTPError: Error {
+        case malformedHead
+        case malformedBody
+        case malformedURL(url: String)
+        case alreadyShutdown
+    }
+
+    /// Initialise HTTPClient
+    public init(eventLoopGroup: NIOTSEventLoopGroup) {
+        self.eventLoopGroup = eventLoopGroup
+    }
+
+    /// send request to HTTP client, return a future holding the Response
+    public func connect(_ request: Request, timeout: TimeAmount) -> EventLoopFuture<Response> {
+        // extract details from request URL
+        guard let url = URL(string:request.head.uri) else { return eventLoopGroup.next().makeFailedFuture(HTTPError.malformedURL(url: request.head.uri)) }
+        guard let scheme = url.scheme else { return eventLoopGroup.next().makeFailedFuture(HTTPError.malformedURL(url: request.head.uri)) }
+        guard let hostname = url.host else { return eventLoopGroup.next().makeFailedFuture(HTTPError.malformedURL(url: request.head.uri)) }
+
+        let port : Int
+        let headerHostname : String
+        if url.port != nil {
+            port = url.port!
+            headerHostname = "\(hostname):\(port)"
+        } else {
+            let isSecure = (scheme == "https")
+            port = isSecure ? 443 : 80
+            headerHostname = hostname
+        }
+
+
+        let response: EventLoopPromise<Response> = self.eventLoopGroup.next().makePromise()
+
+        var bootstrap = NIOTSConnectionBootstrap(group: self.eventLoopGroup)
+            .connectTimeout(timeout)
+            .channelOption(ChannelOptions.socket(IPPROTO_TCP, TCP_NODELAY), value: 1)
+
+        if port == 443 {
+            bootstrap = bootstrap.tlsOptions(NWProtocolTLS.Options())
+        }
+
+        bootstrap.channelInitializer { channel in
+                return channel.pipeline.addHTTPClientHandlers()
+                    .flatMap {
+                        let handlers : [ChannelHandler] = [
+                            HTTPClientRequestSerializer(hostname: headerHostname),
+                            HTTPClientResponseHandler(promise: response)
+                        ]
+                        return channel.pipeline.addHandlers(handlers)
+                }
+            }
+            .connect(host: hostname, port: port)
+            .flatMap { channel -> EventLoopFuture<Void> in
+                return channel.writeAndFlush(request)
+            }
+            .whenFailure { error in
+                response.fail(error)
+        }
```

Comment:
`}` should be indented.

Reply:
whole function was indented weird, fixed